File: buzzvil-sdk-ios/buzzvil-sdk-ios-objc/ViewController.m
#import "ViewController.h"
#import "UIButton+Custom.h"
#import "Native/NativeViewController.h"
#import "Native/Carousel/CarouselViewController.h"
#import "Interstitial/InterstitialViewController.h"
#import "Feed/FeedViewController.h"
#import "Feed/FeedContainerViewController.h"

static NSString * const kNavigationItemTitle = @"BuzzvilSDK-ObjectiveC";
static NSString * const kNativeButtonTitle = @"Native";
static NSString * const kInterstitialButtonTitle = @"Interstitial";
static NSString * const kFeedButtonTitle = @"Feed";
static NSString * const kCarouselButtonTitle = @"Carousel";
static NSString * const kFeedContainerButtonTitle = @"Feed Container";
static CGFloat const kButtonMargin = 12;
static CGFloat const kButtonAspectRatio = 1.5;

@interface ViewController ()

@property (nonatomic, strong, readonly) UIButton *nativeButton;
@property (nonatomic, strong, readonly) UIButton *carouselButton;
@property (nonatomic, strong, readonly) UIButton *interstitialButton;
@property (nonatomic, strong, readonly) UIButton *feedButton;
@property (nonatomic, strong, readonly) UIButton *feedContainerButton;

@end

@implementation ViewController

- (void)viewDidLoad {
  [super viewDidLoad];
  [self setupView];
  [self setupLayout];
  [self setupEvent];
}

#pragma mark - UI setup
- (void)setupView {
  self.navigationItem.title = kNavigationItemTitle;
  
  _nativeButton = [UIButton buttonWithType:UIButtonTypeSystem];
  [self.view addSubview:_nativeButton];
  [_nativeButton setTitle:kNativeButtonTitle forState:UIControlStateNormal];
  [_nativeButton applyCustomStyle];
  
  _interstitialButton = [UIButton buttonWithType:UIButtonTypeSystem];
  [self.view addSubview:_interstitialButton];
  [_interstitialButton setTitle:kInterstitialButtonTitle forState:UIControlStateNormal];
  [_interstitialButton applyCustomStyle];
  
  _carouselButton = [UIButton buttonWithType:UIButtonTypeSystem];
  [self.view addSubview:_carouselButton];
  [_carouselButton setTitle:kCarouselButtonTitle forState:UIControlStateNormal];
  [_carouselButton applyCustomStyle];
  
  
  _feedButton = [UIButton buttonWithType:UIButtonTypeSystem];
  [self.view addSubview:_feedButton];
  [_feedButton setTitle:kFeedButtonTitle forState:UIControlStateNormal];
  [_feedButton applyCustomStyle];
  
  _feedContainerButton = [UIButton buttonWithType:UIButtonTypeSystem];
  [self.view addSubview:_feedContainerButton];
  [_feedContainerButton setTitle:kFeedContainerButtonTitle forState:UIControlStateNormal];
  [_feedContainerButton applyCustomStyle];
}

- (void) setupLayout {
  _nativeButton.translatesAutoresizingMaskIntoConstraints = NO;
  [NSLayoutConstraint activateConstraints:@[
    [_nativeButton.topAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.topAnchor constant:kButtonMargin],
    [_nativeButton.leadingAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.leadingAnchor constant:kButtonMargin],
    [_nativeButton.trailingAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.centerXAnchor constant:-kButtonMargin/2],
    [_nativeButton.widthAnchor constraintEqualToAnchor:_nativeButton.heightAnchor multiplier:kButtonAspectRatio],
  ]];
  
  _interstitialButton.translatesAutoresizingMaskIntoConstraints = NO;
  [NSLayoutConstraint activateConstraints:@[
    [_interstitialButton.topAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.topAnchor constant:kButtonMargin],
    [_interstitialButton.leadingAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.centerXAnchor constant:kButtonMargin/2],
    [_interstitialButton.trailingAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.trailingAnchor constant:-kButtonMargin],
    [_interstitialButton.widthAnchor constraintEqualToAnchor:_interstitialButton.heightAnchor multiplier:kButtonAspectRatio],
  ]];
  
  _carouselButton.translatesAutoresizingMaskIntoConstraints = NO;
  [NSLayoutConstraint activateConstraints:@[
    [_carouselButton.topAnchor constraintEqualToAnchor:_nativeButton.bottomAnchor constant:kButtonMargin],
    [_carouselButton.leadingAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.leadingAnchor constant:kButtonMargin],
    [_carouselButton.trailingAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.centerXAnchor constant:-kButtonMargin / 2],
    [_carouselButton.widthAnchor constraintEqualToAnchor:_carouselButton.heightAnchor multiplier:kButtonAspectRatio],
  ]];
  
  _feedButton.translatesAutoresizingMaskIntoConstraints = NO;
  [NSLayoutConstraint activateConstraints:@[
    [_feedButton.topAnchor constraintEqualToAnchor:_interstitialButton.bottomAnchor constant:kButtonMargin],
    [_feedButton.leadingAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.centerXAnchor constant:kButtonMargin/2],
    [_feedButton.trailingAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.trailingAnchor constant:-kButtonMargin],
    [_feedButton.widthAnchor constraintEqualToAnchor:_feedButton.heightAnchor multiplier:kButtonAspectRatio],
  ]];
  
  _feedContainerButton.translatesAutoresizingMaskIntoConstraints = NO;
  [NSLayoutConstraint activateConstraints:@[
    [_feedContainerButton.topAnchor constraintEqualToAnchor:_carouselButton.bottomAnchor constant:kButtonMargin],
    [_feedContainerButton.leadingAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.leadingAnchor constant:kButtonMargin],
    [_feedContainerButton.trailingAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.centerXAnchor constant:-kButtonMargin / 2],
    [_feedContainerButton.widthAnchor constraintEqualToAnchor:_feedContainerButton.heightAnchor multiplier:kButtonAspectRatio],
  ]];
}


- (void)setupEvent {
  [_feedButton addGestureRecognizer:[[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(pushFeedViewController:)]];
  [_nativeButton addGestureRecognizer:[[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(pushNativeViewController:)]];
  [_carouselButton addGestureRecognizer:[[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(pushCarouselViewController:)]];
  [_interstitialButton addGestureRecognizer:[[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(pushInterstitialViewController:)]];
  [_feedContainerButton addGestureRecognizer:[[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(pushFeedContainerViewController:)]];
}

- (void)pushFeedViewController:(id)sender {
  FeedViewController *feedViewController = [[FeedViewController alloc] init];
  [self.navigationController pushViewController:feedViewController animated:YES];
}

- (void)pushNativeViewController:(id)sender {
  NativeViewController *nativeViewController = [[NativeViewController alloc] init];
  [self.navigationController pushViewController:nativeViewController animated:YES];
}

- (void)pushCarouselViewController:(id)sender {
  CarouselViewController *carouselViewController = [[CarouselViewController alloc] init];
  [self.navigationController pushViewController:carouselViewController animated:YES];
}

- (void)pushInterstitialViewController:(id)sender {
  InterstitialViewController *interstitialViewController = [[InterstitialViewController alloc] init];
  [self.navigationController pushViewController:interstitialViewController animated:YES];
}

- (void)pushFeedContainerViewController:(id)sender {
  FeedContainerViewController *feedContainerViewController = [[FeedContainerViewController alloc] init];
  [self.navigationController pushViewController:feedContainerViewController animated:YES];
}

@end

File: buzzvil-sdk-ios/buzzvil-sdk-ios-swift/ViewController.swift
import UIKit
import BuzzvilSDK

class ViewController: UIViewController {
  private static let navigationItemTitle = "BuzzvilSDK"
  private static let nativeButtonTitle = "Native"
  private static let interstitialButtonTitle = "Interstitial"
  private static let buzzAdFeedButtonTitle = "Feed"
  private static let buzzAdFeedConatinerButtonTitle = "Feed Container"
  private static let carouselButtonTitle = "Carousel"
  private static let inquiryButtonTitle = "Inquiry"
  private static let buttonAspectRatio: CGFloat = 1.5
  private static let buttonMargin: CGFloat = 12
  private static let buttonCornerRadius: CGFloat = 8
  private static let buttonFontSize: CGFloat = 16
  
  private lazy var nativeButton: UIButton = {
    let button = UIButton(frame: .zero)
    button.setTitle(Self.nativeButtonTitle, for: .normal)
    button.addTarget(self, action: #selector(pushNativeAdViewController), for: .touchUpInside)
    return button
  }()
  
  private lazy var interstitialButton: UIButton = {
    let button = UIButton(frame: .zero)
    button.setTitle(Self.interstitialButtonTitle, for: .normal)
    button.addTarget(self, action: #selector(pushInterstitialViewController), for: .touchUpInside)
    return button
  }()
  
  private lazy var carouselButton: UIButton = {
    let button = UIButton(frame: .zero)
    button.setTitle(Self.carouselButtonTitle, for: .normal)
    button.addTarget(self, action: #selector(pushCarouselViewController), for: .touchUpInside)
    return button
  }()
  
  private lazy var feedButton: UIButton = {
    let button = UIButton(frame: .zero)
    button.setTitle(Self.buzzAdFeedButtonTitle, for: .normal)
    button.addTarget(self, action: #selector(pushFeedViewController), for: .touchUpInside)
    return button
  }()
  
  private lazy var feedContainerButton: UIButton = {
    let button = UIButton(frame: .zero)
    button.setTitle(Self.buzzAdFeedConatinerButtonTitle, for: .normal)
    button.addTarget(self, action: #selector(pushFeedContainerViewController), for: .touchUpInside)
    return button
  }()
  
  private lazy var inquiryButton: UIButton = {
    let button = UIButton(frame: .zero)
    button.setTitle(Self.inquiryButtonTitle, for: .normal)
    button.addTarget(self, action: #selector(showInquiry), for: .touchUpInside)
    return button
  }()
  
  override func viewDidLoad() {
    super.viewDidLoad()
    
    setupView()
    setupLayout()
  }
  
  private func setupView() {
    navigationItem.title = Self.navigationItemTitle
    view.addSubview(nativeButton)
    view.addSubview(interstitialButton)
    view.addSubview(carouselButton)
    view.addSubview(feedButton)
    view.addSubview(feedContainerButton)
    
    let buttonsToSetupAppearance = [
      nativeButton,
      interstitialButton,
      carouselButton,
      feedButton,
      feedContainerButton
    ]
    
    buttonsToSetupAppearance.forEach { button in
      setupAppearanceOfButton(button: button)
    }
  }
  
  private func setupLayout() {
    nativeButton.translatesAutoresizingMaskIntoConstraints = false
    NSLayoutConstraint.activate([
      nativeButton.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor, constant: Self.buttonMargin),
      nativeButton.leadingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.leadingAnchor, constant: Self.buttonMargin),
      nativeButton.trailingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.centerXAnchor, constant: -Self.buttonMargin/2),
      nativeButton.widthAnchor.constraint(equalTo: nativeButton.heightAnchor, multiplier: Self.buttonAspectRatio)
    ])
    
    interstitialButton.translatesAutoresizingMaskIntoConstraints = false
    interstitialButton.translatesAutoresizingMaskIntoConstraints = false
    NSLayoutConstraint.activate([
      interstitialButton.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor, constant: Self.buttonMargin),
      interstitialButton.leadingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.centerXAnchor, constant: Self.buttonMargin/2),
      interstitialButton.trailingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.trailingAnchor, constant: -Self.buttonMargin),
      interstitialButton.widthAnchor.constraint(equalTo: interstitialButton.heightAnchor, multiplier: Self.buttonAspectRatio)
    ])
    
    carouselButton.translatesAutoresizingMaskIntoConstraints = false
    NSLayoutConstraint.activate([
      carouselButton.topAnchor.constraint(equalTo: nativeButton.bottomAnchor, constant: Self.buttonMargin),
      carouselButton.leadingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.leadingAnchor, constant: Self.buttonMargin),
      carouselButton.trailingAnchor.constraint(equalTo: view.centerXAnchor, constant: -Self.buttonMargin/2),
      carouselButton.widthAnchor.constraint(equalTo: carouselButton.heightAnchor, multiplier: Self.buttonAspectRatio)
    ])
    
    feedButton.translatesAutoresizingMaskIntoConstraints = false
    NSLayoutConstraint.activate([
      feedButton.topAnchor.constraint(equalTo: interstitialButton.bottomAnchor, constant: Self.buttonMargin),
      feedButton.leadingAnchor.constraint(equalTo: view.centerXAnchor, constant: Self.buttonMargin/2),
      feedButton.trailingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.trailingAnchor, constant: -Self.buttonMargin),
      feedButton.widthAnchor.constraint(equalTo: feedButton.heightAnchor, multiplier: Self.buttonAspectRatio)
    ])
    
    feedContainerButton.translatesAutoresizingMaskIntoConstraints = false
    NSLayoutConstraint.activate([
      feedContainerButton.topAnchor.constraint(equalTo: carouselButton.bottomAnchor, constant: Self.buttonMargin),
      feedContainerButton.leadingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.leadingAnchor, constant: Self.buttonMargin),
      feedContainerButton.trailingAnchor.constraint(equalTo: view.centerXAnchor, constant: -Self.buttonMargin/2),
      feedContainerButton.widthAnchor.constraint(equalTo: feedContainerButton.heightAnchor, multiplier: Self.buttonAspectRatio)
    ])
  }
  
  private func setupAppearanceOfButton(button: UIButton) {
    button.backgroundColor = UIColor(white: 0.15, alpha: 1)
    button.layer.cornerRadius = Self.buttonCornerRadius
    button.setTitleColor(.white, for: .normal)
    button.titleLabel?.font = UIFont.systemFont(ofSize: Self.buttonFontSize)
  }
  
  @objc private func pushNativeAdViewController() {
    let nativeAdViewController = NativeViewController()
    navigationController?.pushViewController(nativeAdViewController, animated: true)
  }
  
  @objc private func pushCarouselViewController() {
    let carouselConfigurationViewController = CarouselViewController()
    navigationController?.pushViewController(carouselConfigurationViewController, animated: true)
  }
  
  @objc private func pushInterstitialViewController() {
    let interstitialViewController = InterstitialViewController()
    navigationController?.pushViewController(interstitialViewController, animated: true)
  }
  
  @objc private func pushFeedViewController() {
    let feedViewController = FeedViewController()
    navigationController?.pushViewController(feedViewController, animated: true)
  }
  
  @objc private func pushFeedContainerViewController() {
    let feedContainerViewController = FeedContainerViewController()
    navigationController?.pushViewController(feedContainerViewController, animated: true)
  }
  
  @objc private func showInquiry() {
    BuzzAdBenefit.presentInquiryPage(on: self, unitId: "YOUR_UNIT_ID")
  }
}
